From 2b40b052c142606b8f44055e7b7ac366b58f0fd0 Mon Sep 17 00:00:00 2001
From: Christian Bruel <christian.bruel@foss.st.com>
Date: Mon, 27 Oct 2025 13:00:10 +0100
Subject: [PATCH] PCI: stm32: enable perst_irq when start link failed

If start link failed, because the host is not started, enable perst_irq to
start link training only after PERST# de-assert.

perst_irq is level triggered, so we can defer link training into the
handler even if the host was already started without enabling
Configuration Request Retry Status.

Change-Id: I761906c9b469fda5f066580ed2858cd395889c7a
Signed-off-by: Christian Bruel <christian.bruel@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/503221
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
Domain-Review: Fabrice GASNIER <fabrice.gasnier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 drivers/pci/controller/dwc/pcie-stm32-ep.c | 107 ++-------------------
 1 file changed, 10 insertions(+), 97 deletions(-)

--- a/drivers/pci/controller/dwc/pcie-stm32-ep.c
+++ b/drivers/pci/controller/dwc/pcie-stm32-ep.c
@@ -18,11 +18,6 @@
 #include "pcie-designware.h"
 #include "pcie-stm32.h"
 
-enum stm32_pcie_ep_link_status {
-	STM32_PCIE_EP_LINK_DISABLED,
-	STM32_PCIE_EP_LINK_ENABLED,
-};
-
 struct stm32_pcie {
 	struct dw_pcie *pci;
 	struct regmap *regmap;
@@ -30,7 +25,6 @@ struct stm32_pcie {
 	struct phy *phy;
 	struct clk *clk;
 	struct gpio_desc *reset_gpio;
-	enum stm32_pcie_ep_link_status link_status;
 	unsigned int perst_irq;
 };
 
@@ -42,69 +36,15 @@ static const struct of_device_id stm32_p
 static void stm32_pcie_ep_init(struct dw_pcie_ep *ep)
 {
 	struct dw_pcie *pci = to_dw_pcie_from_ep(ep);
-	struct stm32_pcie *stm32_pcie = to_stm32_pcie(pci);
 	enum pci_barno bar;
 
 	for (bar = BAR_0; bar <= PCI_STD_NUM_BARS; bar++)
 		dw_pcie_ep_reset_bar(pci, bar);
-
-	/* Defer Completion Requests until link started */
-	regmap_update_bits(stm32_pcie->regmap, SYSCFG_PCIECR,
-			   STM32MP25_PCIECR_REQ_RETRY_EN,
-			   STM32MP25_PCIECR_REQ_RETRY_EN);
-}
-
-static int stm32_pcie_enable_link(struct dw_pcie *pci)
-{
-	struct stm32_pcie *stm32_pcie = to_stm32_pcie(pci);
-	int ret;
-
-	regmap_update_bits(stm32_pcie->regmap, SYSCFG_PCIECR,
-			   STM32MP25_PCIECR_LTSSM_EN,
-			   STM32MP25_PCIECR_LTSSM_EN);
-
-	ret = dw_pcie_wait_for_link(pci);
-	if (ret)
-		return ret;
-
-	regmap_update_bits(stm32_pcie->regmap, SYSCFG_PCIECR,
-			   STM32MP25_PCIECR_REQ_RETRY_EN,
-			   0);
-
-	return 0;
-}
-
-static void stm32_pcie_disable_link(struct dw_pcie *pci)
-{
-	struct stm32_pcie *stm32_pcie = to_stm32_pcie(pci);
-
-	regmap_update_bits(stm32_pcie->regmap, SYSCFG_PCIECR,
-			   STM32MP25_PCIECR_REQ_RETRY_EN,
-			   STM32MP25_PCIECR_REQ_RETRY_EN);
-
-	regmap_update_bits(stm32_pcie->regmap, SYSCFG_PCIECR, STM32MP25_PCIECR_LTSSM_EN, 0);
 }
 
 static int stm32_pcie_start_link(struct dw_pcie *pci)
 {
 	struct stm32_pcie *stm32_pcie = to_stm32_pcie(pci);
-	struct dw_pcie_ep *ep = &pci->ep;
-	int ret;
-
-	if (stm32_pcie->link_status == STM32_PCIE_EP_LINK_ENABLED) {
-		dev_dbg(pci->dev, "Link is already enabled\n");
-		return 0;
-	}
-
-	ret = stm32_pcie_enable_link(pci);
-	if (ret) {
-		dev_err(pci->dev, "PCIe cannot establish link: %d\n", ret);
-		return ret;
-	}
-
-	dw_pcie_ep_linkup(ep);
-
-	stm32_pcie->link_status = STM32_PCIE_EP_LINK_ENABLED;
 
 	enable_irq(stm32_pcie->perst_irq);
 
@@ -115,16 +55,7 @@ static void stm32_pcie_stop_link(struct
 {
 	struct stm32_pcie *stm32_pcie = to_stm32_pcie(pci);
 
-	if (stm32_pcie->link_status == STM32_PCIE_EP_LINK_DISABLED) {
-		dev_dbg(pci->dev, "Link is already disabled\n");
-		return;
-	}
-
 	disable_irq(stm32_pcie->perst_irq);
-
-	stm32_pcie_disable_link(pci);
-
-	stm32_pcie->link_status = STM32_PCIE_EP_LINK_DISABLED;
 }
 
 static int stm32_pcie_raise_irq(struct dw_pcie_ep *ep, u8 func_no,
@@ -198,24 +129,13 @@ static void stm32_pcie_perst_assert(stru
 
 	dev_dbg(dev, "PERST asserted by host. Shutting down the PCIe link\n");
 
-	/*
-	 * Do not try to release resources if the PERST# is
-	 * asserted before the link is started.
-	 */
-	if (stm32_pcie->link_status == STM32_PCIE_EP_LINK_DISABLED) {
-		dev_dbg(pci->dev, "Link is already disabled\n");
-		return;
-	}
-
-	stm32_pcie_disable_link(pci);
+	regmap_update_bits(stm32_pcie->regmap, SYSCFG_PCIECR, STM32MP25_PCIECR_LTSSM_EN, 0);
 
 	dw_pcie_ep_deinit_notify(ep);
 
 	stm32_pcie_disable_resources(stm32_pcie);
 
 	pm_runtime_put_sync(dev);
-
-	stm32_pcie->link_status = STM32_PCIE_EP_LINK_DISABLED;
 }
 
 static void stm32_pcie_perst_deassert(struct dw_pcie *pci)
@@ -225,11 +145,6 @@ static void stm32_pcie_perst_deassert(st
 	struct dw_pcie_ep *ep = &pci->ep;
 	int ret;
 
-	if (stm32_pcie->link_status == STM32_PCIE_EP_LINK_ENABLED) {
-		dev_dbg(pci->dev, "Link is already enabled\n");
-		return;
-	}
-
 	dev_dbg(dev, "PERST de-asserted by host. Starting link training\n");
 
 	ret = pm_runtime_resume_and_get(dev);
@@ -255,16 +170,12 @@ static void stm32_pcie_perst_deassert(st
 
 	dw_pcie_ep_init_notify(ep);
 
-	ret = stm32_pcie_enable_link(pci);
-	if (ret) {
-		dev_err(dev, "PCIe Cannot establish link: %d\n", ret);
-		dw_pcie_ep_deinit_notify(ep);
-		stm32_pcie_disable_resources(stm32_pcie);
-		pm_runtime_put_sync(dev);
-		return;
-	}
+	/* Enable link training */
+	regmap_update_bits(stm32_pcie->regmap, SYSCFG_PCIECR,
+			   STM32MP25_PCIECR_LTSSM_EN,
+			   STM32MP25_PCIECR_LTSSM_EN);
 
-	stm32_pcie->link_status = STM32_PCIE_EP_LINK_ENABLED;
+	dw_pcie_ep_linkup(ep);
 }
 
 static irqreturn_t stm32_pcie_ep_perst_irq_thread(int irq, void *data)
@@ -279,6 +190,9 @@ static irqreturn_t stm32_pcie_ep_perst_i
 	else
 		stm32_pcie_perst_deassert(pci);
 
+	irq_set_irq_type(gpiod_to_irq(stm32_pcie->reset_gpio),
+			 (perst ? IRQF_TRIGGER_HIGH : IRQF_TRIGGER_LOW));
+
 	return IRQ_HANDLED;
 }
 
@@ -396,8 +310,7 @@ static int stm32_pcie_probe(struct platf
 
 	ret = devm_request_threaded_irq(dev, stm32_pcie->perst_irq, NULL,
 					stm32_pcie_ep_perst_irq_thread,
-					IRQF_TRIGGER_RISING |
-					IRQF_TRIGGER_FALLING | IRQF_ONESHOT,
+					IRQF_TRIGGER_HIGH | IRQF_ONESHOT,
 					"perst_irq", stm32_pcie);
 	if (ret) {
 		dev_err(dev, "Failed to request PERST IRQ: %d\n", ret);
